druid-shell/src/platform/x11/window.rs

```diff
@@ -0,0 +1,407 @@
+        };
+        let cairo_surface = cairo::XCBSurface::create(
+            &cairo_xcb_connection,
+            &cairo_drawable,
+            &cairo_visual_type,
+            size.width as i32,
+            size.height as i32,
+        )
+        .expect("couldn't create a cairo surface");
+        let cairo_ctx = cairo::Context::new(&cairo_surface);
+
+        // Figure out the refresh rate of the current screen
+        let refresh_rate = util::refresh_rate(&conn, window_id);
+
+        XWindow {
+            window_id,
+            handler: window_handler,
+            cairo_context: cairo_ctx,
+            refresh_rate,
+        }
+    }
+
+    pub fn render(&mut self) {
+        let conn = Application::get_connection();
+
+        self.cairo_context.set_source_rgb(0.0, 0.0, 0.0);
+        self.cairo_context.paint();
+        let mut piet_ctx = Piet::new(&mut self.cairo_context);
+        let anim = self.handler.paint(&mut piet_ctx);
+        if let Err(e) = piet_ctx.finish() {
+            // TODO(x11/errors): hook up to error or something?
+            panic!("piet error on render: {:?}", e);
+        }
+
+        if anim && self.refresh_rate.is_some() {
+            // TODO(x11/render_improvements): Sleeping is a terrible way to schedule redraws.
+            //     I think I'll end up having to write a redraw scheduler or something. :|
+            //     Doing it this way for now to proof-of-concept it.
+            let sleep_amount_ms = (1000.0 / self.refresh_rate.unwrap()) as u64;
+            std::thread::sleep(std::time::Duration::from_millis(sleep_amount_ms));
+
+            request_redraw(self.window_id);
+        }
+        conn.flush();
+    }
+
+    pub fn key_down(&mut self, key_code: KeyCode) {
+        self.handler.key_down(KeyEvent::new(
+            key_code,
+            false,
+            KeyModifiers {
+                /// Shift.
+                shift: false,
+                /// Option on macOS.
+                alt: false,
+                /// Control.
+                ctrl: false,
+                /// Meta / Windows / Command
+                meta: false,
+            },
+            key_code,
+            key_code,
+        ));
+    }
+
+    pub fn mouse_down(&mut self, mouse_event: &MouseEvent) {
+        self.handler.mouse_down(mouse_event);
+    }
+}
+
+// Apparently you have to get the visualtype this way :|
+fn get_visual_from_screen(screen: &xcb::Screen<'_>) -> Option<xcb::xproto::Visualtype> {
+    for depth in screen.allowed_depths() {
+        for visual in depth.visuals() {
+            if visual.visual_id() == screen.root_visual() {
+                return Some(visual);
+            }
+        }
+    }
+    None
+}
+
+#[derive(Clone)]
+pub struct IdleHandle;
+
+impl IdleHandle {
+    pub fn add_idle_callback<F>(&self, _callback: F)
+    where
+        F: FnOnce(&dyn Any) + Send + 'static,
+    {
+        // TODO(x11/idle_handles): implement IdleHandle::add_idle_callback
+        unimplemented!();
```

Comment:
again I'd just log here instead of panicking

Reply:
Done.